File: m/rls/fidef.r
file	fidef - default file spec
include rid:rider
include rid:chdef
include rid:fidef
include rid:medef
include rid:mxdef
include rid:stdef

;	Handle default filespec
;
;	node:: device: [directory] filename .type ;version
;	device: \directory\directory\ filename .type

data	fiAdef - state table

;	Special treatment for RSX [m,g] specs

  init	fiAdef : [] char
  is	 0 , ':', ':'		; node
	 0 , ':', -1		; device
	'[', ']', -1		; directory
	 0 , '.',  0		; filename
	'.', ';',  0		; filetype
	';',  0,   0		; version
	-1			;
  end

	fi_itm	: (*char, *char) int own
code	fi_def - merge default spec

  func	fi_def
	spc : * char~ 			; file spec
	def : * char			; default spec
	dst : * char			; result spec
	()  : int			; fail => invalid spec
  is	tmp : [mxSPC] char		;
	out : * char~ = tmp		; output pointer
	sta : * char~ = fiAdef		; state table
	lft : int			; spc item length
	rgt : int			; def item length
					;
      repeat				;
	lft = fi_itm (sta, spc)		; parse spec
	rgt = fi_itm (sta, def)		; parse default
	if lft				; found it in spec
	   out = me_cop (spc, out, lft)	; move it in
	else 				; found in default
	.. out = me_cop (def, out, rgt)	; move in default
	*out = 0			; terminate it
	spc += lft			; advance that
	def += rgt			; advance that
	sta += 3			; skip to next set
      until *sta eq -1			; more
	*out = 0			; terminate it
	if (out = st_fnd (":.",tmp)) ne	; got silly result	
	.. out[1] = 0			; device only
					;
	rgt = *spc			; save closing state
	st_cop (tmp, dst)		; return result
					;
	fail if *spc || out eq dst	; invalid filespec
	fine fi_loc (dst, dst)		; fine - convert result
  end
code	fi_itm - get file spec item

  func	fi_itm
	sta : * char~			; state list
	ptr : * char~			; filespec
	()  : int			; item length
  is	str : * char~ = ptr		; string pointer

	fail if str eq NULL		; no string - forget it

;	parse next part
;	check prefix

      repeat
	if *sta++			; need prefix
	.. fail if *(sta-1) ne *str	; wrong prefix

;	Find terminator

	while *str ne *sta		; not at terminator
	   if *str eq 0			; exhausted line
	      if (*sta == '.')		; null matchs dot
              || (*sta == ';')		; null matchs semicolon
	      .. quit			; done
	   .. fail			; else failed
	.. ++str			; next in string

;	Skip or retain terminator

	quit if *++sta eq		; skip terminator
					;
	++str				; retain terminator
	if *sta ne -1			; need double terminator (::)
	   fail if *sta ne *str		; but missed it
	.. ++str			; retain it
      never				;
					;
	reply str-ptr			; return substring length
  end


File: m/rls/rlb/fidef.c
/* file -  fidef - default file spec */
#include "m:\rid\rider.h"
#include "m:\rid\fidef.h"
#include "m:\rid\chdef.h"
#include "m:\rid\medef.h"
#include "m:\rid\mxdef.h"
#include "m:\rid\stdef.h"
/* code -  fiAdef - state table */
char fiAdef [] =  {
  0 , ':', ':',0 , ':', -1,
  '[', ']', -1,
  0 , '.', 0,
  '.', ';', 0,
  ';', 0, 0,
  -1,
  };
static int fi_itm (char *,char *);
/* code -  fi_def - merge default spec */
int fi_def(
char *src ,
char *def ,
char *dst )
{ char buf [mxSPC];
  char *spc = buf;
  register int sps ;
  register int dfs ;
  register char *sta = fiAdef;
  register char *out = dst;
  st_cop (src, spc);
  for(;;)  {
    sps = fi_itm (sta, spc);
    dfs = fi_itm (sta, def);
    if ( sps) {
      out = me_cop (spc, out, sps);
      } else {
      out = me_cop (def, out, dfs); }
    *out = 0;
    spc += sps;
    def += dfs;
    sta += 3;
  if( *sta == -1)break; }
  *out = 0;
  if( *spc || out == dst)return 0;
   fi_loc (dst, dst);return 1;
} 
/* code -  fi_itm - get file spec item */
int fi_itm(
register char *sta ,
register char *ptr )
{ register char *str = ptr;
  if( str == NULL)return 0;
  for(;;)  {
    if ( *sta++) {
      if( *(sta-1) != *str)return 0; }
    while ( *str != *sta) {
      if ( *str == 0) {
        if (( (*sta == '.'))
        ||((*sta == ';'))) {
          break; }
        return 0; }
      ++str; }
    if( *++sta == 0)break;
    ++str;
    if ( *sta != -1) {
      if( *sta != *str)return 0;
      ++str; }
   break;} 
  return ( str-ptr);
} 
